### PR TITLE
Implement textDocument/selectionRange

### LINKED ISSUE #100: Implement textDocument/selectionRange
The most recent version of the specification (3.15) introduced a `textDocument/selectionRange` request. This request is supposed to suggest available selections to the server.

This very closely corresponds to the `Shape` command in merlin (cf `query_protocol.ml`). There is just a bit of a mismatch because merlin returns a tree of shapes (every shape can have many sub shapes), so we'll need to do some intelligent flattening. This last bit we can tweak experimentally after trying out an initial implementation.

### PR DESCRIPTION
**Pull request by cannorin**
Closes #100.

This PR implements `textDocument/selectionRange`.

This is still WIP:
- [x] Find the desired way to flatten Merlin's `Shape` into LSP's `SelectionRange`
- [x] Change the server's capabilities
- [x] Write some tests

### CODE REVIEW

File: ocaml-lsp-server/test/e2e/__tests__/textDocument-formatting.test.ts [hunk 1]
```diff
@@ -108,13 +110,15 @@ describe("textDocument/formatting", () => {
...
-            "end", ""].join(newline)
+        newText: [
+          "module Test : sig",
+          "  type t =",
+          "    | Foo",
+          "    | Bar",
+          "    | Baz",
+          "end",
+          "",
+        ].join(newline),
```
Comment (rgrinberg) [inline ocaml-lsp-server/test/e2e/__tests__/textDocument-formatting.test.ts]: Why do we need these unrelated changes? Perhaps they can be in a separate PR?

Comment (cannorin) [reply ocaml-lsp-server/test/e2e/__tests__/textDocument-formatting.test.ts]: that was because of running `make fmt`. I will exclude this change.

Comment (rgrinberg) [reply ocaml-lsp-server/test/e2e/__tests__/textDocument-formatting.test.ts]: Yes it because of the recent PR that changed the ocamlformat tests? If so we can revert those changes.

Comment (rgrinberg) [reply ocaml-lsp-server/test/e2e/__tests__/textDocument-formatting.test.ts]: I reverterd those changes in master. So if you rebase, you shouldn't need these unrelated changes anymore.

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 2]
```diff
@@ -683,7 +684,52 @@ let on_request :
...
+            )
+      in
+      (* try to find the nearest range inside first, then outside *)
+      let nearest_range =
+        List.map ranges ~f:(fun r ->
+            ( r
+            , range_includes_pos r.Lsp.Protocol.SelectionRange.range
+                cursor_position ))
+        |> List.sort ~compare:(fun r1 r2 ->
+               compare (snd r1) (snd r2) |> Ordering.of_int)
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/ocaml_lsp_server.ml]: You're comparing a polymorphic variant using polymorphic comparison. This is quite suspicious to me and we should write an explicit comparison function here.

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 3]
```diff
@@ -696,28 +696,37 @@ let on_request :
...
+              Ordering.of_int (l1 - l2)
+        in
         List.map ranges ~f:(fun r ->
             ( r
             , range_includes_pos r.Lsp.Protocol.SelectionRange.range
                 cursor_position ))
-        |> List.sort ~compare:(fun r1 r2 ->
-               compare (snd r1) (snd r2) |> Ordering.of_int)
+        |> List.sort ~compare:(fun r1 r2 -> compare_inclusion (snd r1) (snd r2))
         |> List.map ~f:fst |> List.hd_opt
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/ocaml_lsp_server.ml]: This seems like you're just finding the minimum. How about we just do it in one pass without sorting/mapping.

We can start by writing `min_by : 'a list -> f:('a -> 'a -> Ordering.t) -> 'a option`

And then simply call it: `min_by ranges ~f:(fun x y -> let on r = range_includes_pos .. in compare_inclusion (on x) (on y))`.

What do you think?

Comment (cannorin) [reply ocaml-lsp-server/src/ocaml_lsp_server.ml]: sounds good.

**Comment by rgrinberg** [comment]
I've pushed a commit that moves the range math outside the main function. This will make it easier for rebase my generated code changes. Also, I've cleaned up the code as I was trying to understand it. Take a look to see if you agree with the changes.

**Comment by cannorin** [comment]
Looks good!
